Sources/ComposableArchitecture/Instrumentation.swift

```diff
@@ -0,0 +1,142 @@
+public class Instrumentation {
+  public typealias Trigger = (EventInfo) -> Void
+
+  /// Container for the information that will be provided to tracking/instrumentation implementations.
+  public struct EventInfo: CustomStringConvertible {
+    internal init(type: String, action: String = "", tags: [String: String] = [:]) {
+      self.type = type
+      self.action = action
+      self.tags = tags
+    }
+
+    /// The Swift type of object that is operating. This will generally be of the type `Store<State, Action>` or
+    /// `ViewStore<State, Action>` with the `State` and `Action` types properly filled in. With this information it
+    /// _should_ be possible to identify which ``Store`` or ``ViewStore`` (or other type) is operating.
+    public let type: String
+
+    /// A ``String`` generated for the ``Action`` that was sent, when available. There are operations that may not have
+    /// an ``Action`` available and so this may be an empty string. The value is generated using ``String(describing:)``
+    /// and so is dependent on the Swift runtime metadata. If that metadata is removed in some way then this value will
+    /// be empty most likely.
+    public var action: String
+
+    /// A dictionary of tags that the library thinks are valuable to include in the tracking/instrumentation. There is
+    /// obviously no requirement to use these, but they are there just in case.
+    public var tags: [String: String]
+
+    public var description: String {
+      guard !action.isEmpty else {
+        return "\(type)"
+      }
+
+      return "\(type): \(action)"
+    }
+  }
+
+  public init(viewStore: Instrumentation.ViewStore? = nil, store: Instrumentation.Store? = nil) {
+    self.viewStore = viewStore
+    self.store = store
+  }
+
+  public static var shared: Instrumentation = .noop
+
+  /// Tracking/instrumentation hooks that operate only within the context of ``ViewStore`` objects.
+  public struct ViewStore {
+    public init(willSend: @escaping Trigger, didSend: @escaping Trigger, willDeduplicate: @escaping Trigger, didDeduplicate: @escaping Trigger, stateWillChange: @escaping Trigger, stateDidChange: @escaping Trigger) {
+      self.willSend = willSend
+      self.didSend = didSend
+      self.willDeduplicate = willDeduplicate
+      self.didDeduplicate = didDeduplicate
+      self.stateWillChange = stateWillChange
+      self.stateDidChange = stateDidChange
+    }
+
+    /// Called _before_ the ``ViewStore.send`` handles the action.
+    let willSend: Trigger
+    /// Called  _after_ the ``ViewStore.send`` has completed handling the action.
+    let didSend: Trigger
+    /// Called _before_ the ``ViewStore`` attempts to deduplicate the old and new states. It is expected that for every
+    /// ``willDeduplicate`` there will be a matching ``didDeduplicate``.
+    /// Note: This may _not_ be called in every case. Because the deduplication implementation uses the
```

Comment:
l.91-93 seem to be unnecessarily vague; is it correct to instead say "Not triggered for the first state value."?